RavenDB-18873 JavaScript projections from Corax indexes.

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18873

### Additional description

Projection from index via RQL or JavaScript for Corax indexes.

### Type of change


- New feature

### How risky is the change?

- High

### Backward compatibility


- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works


### Testing by RavenDB QA team


- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- We've to revert `Store` options modification we did for Corax.

maciejaszyk committed:
RavenDB-18873 Adding tests for projections with Corax

maciejaszyk committed:
RavenDB-18873 Increment skippedResults when nothing is projected.

maciejaszyk committed:
RavenDB-18873 In case of empty or null string we've to translate the original term in builder.

maciejaszyk committed:
RavenDB-18873 Projections from Corax JS index.

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
+                ref var reader = ref parent.IndexRetriever.CoraxEntry;
+                var fieldReader = binding is not null
+                    ? reader.GetFieldReaderFor(binding.Metadata)
+                    : reader.GetFieldReaderFor(Encodings.Utf8.GetBytes(property));
```
Do you need to call `Encodings.Utf8.GetBytes` every time? Maybe cache it?

maciejaszyk replied:
Done

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
+                            }
+                            else if (iterator.IsEmptyCollection)
+                            {
+                                throw new InvalidDataException("Tuple list cannot contain an empty string (otherwise, where did the numeric came from!)");
```
Maybe it's worth do add more debug details here e.g. `property` value?

arekpalinski replied:
Reminder about that

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
 
-                        value = FromObject(parent.Engine, arrayItems);
-                        return true;
+                        if (valueInEntry.Contains((byte)'.'))
```
Do you need to check it this way? Cannot `TryReadTuple` return whether it read double or long?

maciejaszyk replied:
No, the tuple always has [string, double, long] value inside.

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
+                            fixed (byte* ptr = &iterator.Sequence.GetPinnableReference())
+                            {
+                                var itemAsBlittable = new BlittableJsonReaderObject(ptr, iterator.Sequence.Length, parent.Blittable._context);
+                                arrayItems[idX] = TranslateToJs(parent, indexField.Name, BlittableJsonToken.StartObject, itemAsBlittable);
```
Is it okay that `ptr` will no longer be pinned? Can an item in `arrayItems` still reference that pointer?

maciejaszyk replied:
@redknightlois  when we take an entry from the container, the pointer is valid until we close the transaction, right?

arekpalinski replied:
If that was the case then why we'd need to wrap it by `fixed` ?

redknightlois replied:
Unless we can guarantee that the content will always be a pointer, the answer is no.

maciejaszyk replied:
We make a clone inside `TranslateToJs`:
```csharp
 if (o is BlittableJsonReaderObject json)
{
    return new BlittableObjectInstance(engine, null, Clone(json, context), null, null, null);
}
  ```

arekpalinski replied:
okay

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
+                        fixed (byte* ptr = &blittableBinary.GetPinnableReference())
+                        {
+                            var itemAsBlittable = new BlittableJsonReaderObject(ptr, blittableBinary.Length, parent.Blittable._context);
+                            value = TranslateToJs(parent, indexField.Name, BlittableJsonToken.StartObject, itemAsBlittable);
```
Same question as above

maciejaszyk replied:
answered above

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
+                    Lucene.Net.Documents.Field[] propertyFields = parent.IndexRetriever.LuceneDocument.GetFields(property);
+
+                    JsValue[] arrayItems =
+                        new JsValue[propertyFields.Length];
```
Formatting

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -138,410 +144,589 @@ private bool TryGetValueFromDocument(BlittableObjectInstance parent, string prop
-                        _parent.MarkChanged();
-                        BlittableJsonReaderObject blittable = (BlittableJsonReaderObject)value;
+                case BlittableJsonToken.Integer:
+                    // TODO: in the future, add [numeric type]TryFormat, when parsing numbers to strings
```
Let's talk about this TODO directly. I don't understand

maciejaszyk replied:
It's not my comment, something happened with the formatting and it appeared as a new one. Sorry.

maciejaszyk replied:
Fixed, now only my changes are visible.

maciejaszyk committed:
RavenDB-18873 In case when field doesn't exist in index let's skip sorting.

maciejaszyk committed:
RavenDB-18873
- allow to project id() from index
- support of dynamic field JS projections
- add string as key to dictionary at `IndexFieldsMapping`
- support for Time projection in case of Tuple (when we store Ticks)

maciejaszyk committed:
RavenDB-18873 Throw when field is not stored nor indexed.

maciejaszyk committed:
RavenDB-18873 Adjusting tests for Corax ( since we've different behaviour than Lucene).

maciejaszyk committed:
RavenDB-7070 Cleanup of spatial query (corax)

maciejaszyk committed:
RavenDB-18873: For field that is not stored or indexed: check should only happen during indexing.

maciejaszyk committed:
RavenDB-18873 More tests for JS projections & optimatizations & nullable numeric lists - writer fix.

maciejaszyk committed:
RavenDB-18873 Caching dynamic field property name as Slice in Corax's IndexSearcher.

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -121,6 +124,171 @@ public BlittableObjectProperty(BlittableObjectInstance parent, string property)
+                if (parent.IndexFields.TryGetValue(_property, out var indexField) == false && parent.AnyDynamicIndexFields == false)
+                    return false;
+
+                bool isLucene = parent.IndexRetriever.LuceneDocument != null && parent.IndexRetriever.State != null;
```
Isn't the first condition `parent.IndexRetriever.LuceneDocument != null` enough?

arekpalinski replied:
Actually, in old code we had a bit different conditions:

https://github.com/ravendb/ravendb/pull/16088/files#diff-2e22ccd65749e9ac7c17344149bff4b1cb1c8daa40edebd3c4f2d7f83c67769dL148

arekpalinski commented on src/Raven.Server/Documents/Patch/BlittableObjectInstance.cs:
```diff
@@ -121,6 +124,171 @@ public BlittableObjectProperty(BlittableObjectInstance parent, string property)
+                        var hasTime = parent.IndexRetriever.IndexFieldsPersistence.HasTimeValues(binding?.FieldNameAsString ?? property);
+                        if (hasTime)
+                            value = Encodings.Utf8.GetString(valueInEntry);
+                        else if (valueInEntry.Contains((byte)'.'))
```
Can we detect if that's long by checking if `dVal.AlmostEquals(lVal)`? If that's not the case then we know it's double

arekpalinski commented on src/Raven.Server/Documents/Queries/QueryBuilderHelper.cs:
```diff
@@ -528,9 +528,10 @@ internal static QueryFieldName ExtractIndexFieldName(Query query, BlittableJsonR
-            || fieldName is Constants.Documents.Indexing.Fields.DocumentIdFieldName)
+        
+        //Sometimes index can contians Id property and its different than real document ID. We've to 
+        if ((fieldName.Equals(Constants.Documents.Indexing.Fields.DocumentIdMethodName, StringComparison.OrdinalIgnoreCase) && indexMapping.ContainsField(fieldName)) == false && 
```
You changed the operator and now instead of `||` you have `&&` between conditions. Can we have a test for that case?

maciejaszyk replied:
Done

arekpalinski commented on src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs:
```diff
@@ -158,12 +161,12 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
                 {
-
-                    if (FieldsToFetch.ExtractAllFromIndex)
+                    if (retrieverInput is not {LuceneDocument: null, State: null})
```
Same questions about the condition like above. Maybe we should have some method like `IsLuceneDocument()`?

maciejaszyk replied:
Good idea! Done

maciejaszyk committed:
RavenDB-18873:
- Introducing IsLuceneDocument() method in RetrieverInput.
- Changing the check for double from detecting a dot to calculating the epsilon between two numbers (double/long).

maciejaszyk committed:
RavenDB-18873 Tests for query difference between `id()` and `id`
